### PR TITLE
version: extend token comparisons to strings, integers, and nil

### PR DESCRIPTION
**Pull request by SeekingMeaning**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This is useful for comparing version tokens, especially with #8183, for example:

```rb
Formula["gcc"].version.major > 6
```

### CODE REVIEW

File: Library/Homebrew/version.rb [hunk 1]
```diff
@@ -465,17 +508,7 @@ def max(a, b)
...
-      when /\A#{AlphaToken::PATTERN}\z/o   then AlphaToken
-      when /\A#{BetaToken::PATTERN}\z/o    then BetaToken
-      when /\A#{RCToken::PATTERN}\z/o      then RCToken
-      when /\A#{PreToken::PATTERN}\z/o     then PreToken
-      when /\A#{PatchToken::PATTERN}\z/o   then PatchToken
-      when /\A#{NumericToken::PATTERN}\z/o then NumericToken
-      when /\A#{StringToken::PATTERN}\z/o  then StringToken
-      end.new(token)
-    end
+    version.scan(SCAN_PATTERN).map! { |token| Token.create(token) }
```
Comment (reitermarkus) [inline Library/Homebrew/version.rb:511]:
```suggestion
    version.scan(SCAN_PATTERN).map { |token| Token.create(token) }
```

File: Library/Homebrew/version.rb [hunk 2]
```diff
@@ -37,6 +61,8 @@ def initialize(value = nil)
     end
 
     def <=>(other)
+      return unless other = Token.from(other)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/version.rb:64]: Should this be returning `nil` or `-1`? It feels like this should maybe be part of the else case?

Comment (Bo98) [reply Library/Homebrew/version.rb:64]: `nil` represents that the two types are not comparable. In this case, if `other` isn't a valid token (and cannot be converted to one).

This line will also do a conversion from int/string to a token before hitting the `case` checks.

Comment (MikeMcQuaid) [reply Library/Homebrew/version.rb:64]: Works for me, thanks @Bo98!

**Pull request merged**